File: maps/map/src/config/DevConfig.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__New = ____lualib.__TS__New
local __TS__StringSplit = ____lualib.__TS__StringSplit
local __TS__NumberToFixed = ____lualib.__TS__NumberToFixed
local __TS__StringPadStart = ____lualib.__TS__StringPadStart
local __TS__ArrayIndexOf = ____lualib.__TS__ArrayIndexOf
local __TS__ObjectAssign = ____lualib.__TS__ObjectAssign
local ____exports = {}
--- 默认开发配置
____exports.DEFAULT_DEV_CONFIG = {hotReload = {enabled = true, autoStart = true, checkInterval = 2, preserveState = true}, debug = {enabled = true, logLevel = "info", showTimestamps = true}, commands = {enabled = true, adminOnly = true}}
--- 开发环境管理器
____exports.DevEnvironment = __TS__Class()
local DevEnvironment = ____exports.DevEnvironment
DevEnvironment.name = "DevEnvironment"
function DevEnvironment.prototype.____constructor(self, config)
    self.config = config
    self.startTime = os.time()
end
function DevEnvironment.getInstance(self, config)
    if not ____exports.DevEnvironment.instance then
        ____exports.DevEnvironment.instance = __TS__New(____exports.DevEnvironment, config or ____exports.DEFAULT_DEV_CONFIG)
    end
    return ____exports.DevEnvironment.instance
end
function DevEnvironment.prototype.initialize(self)
    self:log("info", "=== 开发环境初始化开始 ===")
    self:printConfig()
    if self.config.debug.enabled then
        self:setupDebugEnvironment()
    end
    if self.config.commands.enabled then
        self:setupDevelopmentCommands()
    end
    self:log("info", "=== 开发环境初始化完成 ===")
end
function DevEnvironment.prototype.printConfig(self)
    self:log("info", "开发环境配置:")
    self:log("info", "  热更新: " .. (self.config.hotReload.enabled and "启用" or "禁用"))
    self:log("info", "  调试模式: " .. (self.config.debug.enabled and "启用" or "禁用"))
    self:log("info", "  开发命令: " .. (self.config.commands.enabled and "启用" or "禁用"))
end
function DevEnvironment.prototype.setupDebugEnvironment(self)
    if self.config.debug.showTimestamps then
        local originalPrint = print
        _G.print = function(message)
            local timestamp = self:getFormattedTime()
            originalPrint((("[" .. timestamp) .. "] ") .. message)
        end
    end
    self:log("info", "调试环境已设置")
end
function DevEnvironment.prototype.setupDevelopmentCommands(self)
    local trigger = CreateTrigger()
    do
        local i = 0
        while i < 8 do
            TriggerRegisterPlayerChatEvent(
                trigger,
                Player(i),
                "-dev",
                false
            )
            i = i + 1
        end
    end
    TriggerAddAction(
        trigger,
        function()
            local chatString = GetEventPlayerChatString()
            local player = GetTriggerPlayer()
            if self.config.commands.adminOnly and GetPlayerId(player) ~= 0 then
                return
            end
            self:handleDevCommand(chatString)
        end
    )
    self:log("info", "开发命令已注册")
    self:log("info", "可用命令: -dev help")
end
function DevEnvironment.prototype.handleDevCommand(self, command)
    local parts = __TS__StringSplit(command, " ")
    local cmd = parts[2] or "help"
    repeat
        local ____switch18 = cmd
        local ____cond18 = ____switch18 == "help"
        if ____cond18 then
            self:showHelpMenu()
            break
        end
        ____cond18 = ____cond18 or ____switch18 == "status"
        if ____cond18 then
            self:showStatus()
            break
        end
        ____cond18 = ____cond18 or ____switch18 == "config"
        if ____cond18 then
            self:showCurrentConfig()
            break
        end
        ____cond18 = ____cond18 or ____switch18 == "performance"
        if ____cond18 then
            self:showPerformanceInfo()
            break
        end
        ____cond18 = ____cond18 or ____switch18 == "memory"
        if ____cond18 then
            self:showMemoryInfo()
            break
        end
        ____cond18 = ____cond18 or ____switch18 == "time"
        if ____cond18 then
            self:showTimeInfo()
            break
        end
        do
            self:log("warn", ("未知命令: " .. cmd) .. "。输入 -dev help 查看帮助")
            break
        end
    until true
end
function DevEnvironment.prototype.showHelpMenu(self)
    self:log("info", "=== 开发命令帮助 ===")
    self:log("info", "-dev help        - 显示此帮助")
    self:log("info", "-dev status      - 显示系统状态")
    self:log("info", "-dev config      - 显示当前配置")
    self:log("info", "-dev performance - 显示性能信息")
    self:log("info", "-dev memory      - 显示内存信息")
    self:log("info", "-dev time        - 显示时间信息")
end
function DevEnvironment.prototype.showStatus(self)
    local uptime = os.time() - self.startTime
    self:log("info", "=== 系统状态 ===")
    self:log(
        "info",
        ("运行时间: " .. tostring(uptime)) .. " 秒"
    )
    self:log("info", "热更新: " .. (self.config.hotReload.enabled and "活跃" or "未启用"))
    self:log("info", "调试模式: " .. (self.config.debug.enabled and "开启" or "关闭"))
end
function DevEnvironment.prototype.showCurrentConfig(self)
    self:log("info", "=== 当前配置 ===")
    self:log(
        "info",
        JSON:stringify(self.config, nil, 2)
    )
end
function DevEnvironment.prototype.showPerformanceInfo(self)
    self:log("info", "=== 性能信息 ===")
    local memoryKB = collectgarbage("count")
    self:log(
        "info",
        ("Lua 内存使用: " .. __TS__NumberToFixed(memoryKB, 2)) .. " KB"
    )
    self:log(
        "info",
        ("FPS: " .. GetLocalizedString("GAMEERR_FRAME_RATE")) .. " (模拟)"
    )
end
function DevEnvironment.prototype.showMemoryInfo(self)
    self:log("info", "=== 内存信息 ===")
    local beforeGC = collectgarbage("count")
    collectgarbage("collect")
    local afterGC = collectgarbage("count")
    self:log(
        "info",
        ("GC 前: " .. __TS__NumberToFixed(beforeGC, 2)) .. " KB"
    )
    self:log(
        "info",
        ("GC 后: " .. __TS__NumberToFixed(afterGC, 2)) .. " KB"
    )
    self:log(
        "info",
        ("释放: " .. __TS__NumberToFixed(beforeGC - afterGC, 2)) .. " KB"
    )
end
function DevEnvironment.prototype.showTimeInfo(self)
    local current = os.time()
    local uptime = current - self.startTime
    self:log("info", "=== 时间信息 ===")
    self:log(
        "info",
        "当前时间戳: " .. tostring(current)
    )
    self:log(
        "info",
        "启动时间戳: " .. tostring(self.startTime)
    )
    self:log(
        "info",
        ("运行时长: " .. tostring(uptime)) .. " 秒"
    )
    self:log(
        "info",
        "格式化时间: " .. self:getFormattedTime()
    )
end
function DevEnvironment.prototype.getFormattedTime(self)
    local time = os.time()
    local minutes = math.floor(time / 60) % 60
    local seconds = time % 60
    return (__TS__StringPadStart(
        tostring(minutes),
        2,
        "0"
    ) .. ":") .. __TS__StringPadStart(
        tostring(seconds),
        2,
        "0"
    )
end
function DevEnvironment.prototype.log(self, level, message)
    if self:shouldLog(level) then
        local prefix = self:getLogPrefix(level)
        print(prefix .. message)
    end
end
function DevEnvironment.prototype.shouldLog(self, level)
    local levels = {"info", "warn", "error"}
    local currentLevelIndex = __TS__ArrayIndexOf(levels, self.config.debug.logLevel)
    local messageLevelIndex = __TS__ArrayIndexOf(levels, level)
    return messageLevelIndex >= currentLevelIndex
end
function DevEnvironment.prototype.getLogPrefix(self, level)
    local prefixes = {info = "[INFO] ", warn = "[WARN] ", error = "[ERROR] "}
    return prefixes[level]
end
function DevEnvironment.prototype.updateConfig(self, newConfig)
    self.config = __TS__ObjectAssign({}, self.config, newConfig)
    self:log("info", "开发配置已更新")
end
function DevEnvironment.prototype.getConfig(self)
    return __TS__ObjectAssign({}, self.config)
end
return ____exports

File: src/config/DevConfig.ts
/** @noSelfInFile */

/**
 * 开发环境配置
 * 管理热更新和开发工具的设置
 */

export interface DevConfig {
  /** 是否启用热更新 */
  hotReload: {
    enabled: boolean;
    autoStart: boolean;
    checkInterval: number;
    preserveState: boolean;
  };
  
  /** 调试设置 */
  debug: {
    enabled: boolean;
    logLevel: 'info' | 'warn' | 'error';
    showTimestamps: boolean;
  };
  
  /** 开发命令 */
  commands: {
    enabled: boolean;
    adminOnly: boolean;
  };
}

/**
 * 默认开发配置
 */
export const DEFAULT_DEV_CONFIG: DevConfig = {
  hotReload: {
    enabled: true,        // 启用热更新
    autoStart: true,      // 自动启动
    checkInterval: 2,     // 2秒检查间隔
    preserveState: true   // 保持状态
  },
  
  debug: {
    enabled: true,        // 启用调试
    logLevel: 'info',     // 信息级别日志
    showTimestamps: true  // 显示时间戳
  },
  
  commands: {
    enabled: true,        // 启用开发命令
    adminOnly: true       // 仅管理员可用
  }
};

/**
 * 开发环境管理器
 */
export class DevEnvironment {
  private static instance: DevEnvironment;
  private config: DevConfig;
  private startTime: number;

  private constructor(config: DevConfig) {
    this.config = config;
    this.startTime = os.time();
  }

  public static getInstance(config?: DevConfig): DevEnvironment {
    if (!DevEnvironment.instance) {
      DevEnvironment.instance = new DevEnvironment(config || DEFAULT_DEV_CONFIG);
    }
    return DevEnvironment.instance;
  }

  /**
   * 初始化开发环境
   */
  public initialize(): void {
    this.log('info', '=== 开发环境初始化开始 ===');
    
    // 显示配置信息
    this.printConfig();
    
    // 设置调试环境
    if (this.config.debug.enabled) {
      this.setupDebugEnvironment();
    }
    
    // 设置开发命令
    if (this.config.commands.enabled) {
      this.setupDevelopmentCommands();
    }
    
    this.log('info', '=== 开发环境初始化完成 ===');
  }

  /**
   * 打印配置信息
   */
  private printConfig(): void {
    this.log('info', '开发环境配置:');
    this.log('info', `  热更新: ${this.config.hotReload.enabled ? '启用' : '禁用'}`);
    this.log('info', `  调试模式: ${this.config.debug.enabled ? '启用' : '禁用'}`);
    this.log('info', `  开发命令: ${this.config.commands.enabled ? '启用' : '禁用'}`);
  }

  /**
   * 设置调试环境
   */
  private setupDebugEnvironment(): void {
    // 重定向 print 函数以添加时间戳
    if (this.config.debug.showTimestamps) {
      const originalPrint = print;
      
      // @ts-ignore
      _G.print = (message: string) => {
        const timestamp = this.getFormattedTime();
        originalPrint(`[${timestamp}] ${message}`);
      };
    }
    
    this.log('info', '调试环境已设置');
  }

  /**
   * 设置开发命令
   */
  private setupDevelopmentCommands(): void {
    const trigger = CreateTrigger();
    
    // 注册聊天事件
    for (let i = 0; i < 8; i++) {
      TriggerRegisterPlayerChatEvent(trigger, Player(i), "-dev", false);
    }
    
    TriggerAddAction(trigger, () => {
      const chatString = GetEventPlayerChatString();
      const player = GetTriggerPlayer();
      
      // 权限检查
      if (this.config.commands.adminOnly && GetPlayerId(player) !== 0) {
        return;
      }
      
      this.handleDevCommand(chatString);
    });
    
    this.log('info', '开发命令已注册');
    this.log('info', '可用命令: -dev help');
  }

  /**
   * 处理开发命令
   */
  private handleDevCommand(command: string): void {
    const parts = command.split(' ');
    const cmd = parts[1] || 'help';
    
    switch (cmd) {
      case 'help':
        this.showHelpMenu();
        break;
        
      case 'status':
        this.showStatus();
        break;
        
      case 'config':
        this.showCurrentConfig();
        break;
        
      case 'performance':
        this.showPerformanceInfo();
        break;
        
      case 'memory':
        this.showMemoryInfo();
        break;
        
      case 'time':
        this.showTimeInfo();
        break;
        
      default:
        this.log('warn', `未知命令: ${cmd}。输入 -dev help 查看帮助`);
        break;
    }
  }

  /**
   * 显示帮助菜单
   */
  private showHelpMenu(): void {
    this.log('info', '=== 开发命令帮助 ===');
    this.log('info', '-dev help        - 显示此帮助');
    this.log('info', '-dev status      - 显示系统状态');
    this.log('info', '-dev config      - 显示当前配置');
    this.log('info', '-dev performance - 显示性能信息');
    this.log('info', '-dev memory      - 显示内存信息');
    this.log('info', '-dev time        - 显示时间信息');
  }

  /**
   * 显示系统状态
   */
  private showStatus(): void {
    const uptime = os.time() - this.startTime;
    this.log('info', '=== 系统状态 ===');
    this.log('info', `运行时间: ${uptime} 秒`);
    this.log('info', `热更新: ${this.config.hotReload.enabled ? '活跃' : '未启用'}`);
    this.log('info', `调试模式: ${this.config.debug.enabled ? '开启' : '关闭'}`);
  }

  /**
   * 显示当前配置
   */
  private showCurrentConfig(): void {
    this.log('info', '=== 当前配置 ===');
    this.log('info', JSON.stringify(this.config, null, 2));
  }

  /**
   * 显示性能信息
   */
  private showPerformanceInfo(): void {
    this.log('info', '=== 性能信息 ===');
    
    // Lua 的垃圾回收信息
    const memoryKB = collectgarbage('count');
    this.log('info', `Lua 内存使用: ${memoryKB.toFixed(2)} KB`);
    
    // 其他性能指标
    this.log('info', `FPS: ${GetLocalizedString("GAMEERR_FRAME_RATE")} (模拟)`);
  }

  /**
   * 显示内存信息
   */
  private showMemoryInfo(): void {
    this.log('info', '=== 内存信息 ===');
    
    const beforeGC = collectgarbage('count');
    collectgarbage('collect');
    const afterGC = collectgarbage('count');
    
    this.log('info', `GC 前: ${beforeGC.toFixed(2)} KB`);
    this.log('info', `GC 后: ${afterGC.toFixed(2)} KB`);
    this.log('info', `释放: ${(beforeGC - afterGC).toFixed(2)} KB`);
  }

  /**
   * 显示时间信息
   */
  private showTimeInfo(): void {
    const current = os.time();
    const uptime = current - this.startTime;
    
    this.log('info', '=== 时间信息 ===');
    this.log('info', `当前时间戳: ${current}`);
    this.log('info', `启动时间戳: ${this.startTime}`);
    this.log('info', `运行时长: ${uptime} 秒`);
    this.log('info', `格式化时间: ${this.getFormattedTime()}`);
  }

  /**
   * 获取格式化时间
   */
  private getFormattedTime(): string {
    const time = os.time();
    const minutes = Math.floor(time / 60) % 60;
    const seconds = time % 60;
    return `${String(minutes).padStart(2, '0')}:${String(seconds).padStart(2, '0')}`;
  }

  /**
   * 日志函数
   */
  private log(level: 'info' | 'warn' | 'error', message: string): void {
    if (this.shouldLog(level)) {
      const prefix = this.getLogPrefix(level);
      print(`${prefix}${message}`);
    }
  }

  /**
   * 检查是否应该记录日志
   */
  private shouldLog(level: 'info' | 'warn' | 'error'): boolean {
    const levels = ['info', 'warn', 'error'];
    const currentLevelIndex = levels.indexOf(this.config.debug.logLevel);
    const messageLevelIndex = levels.indexOf(level);
    
    return messageLevelIndex >= currentLevelIndex;
  }

  /**
   * 获取日志前缀
   */
  private getLogPrefix(level: 'info' | 'warn' | 'error'): string {
    const prefixes = {
      info: '[INFO] ',
      warn: '[WARN] ',
      error: '[ERROR] '
    };
    
    return prefixes[level];
  }

  /**
   * 更新配置
   */
  public updateConfig(newConfig: Partial<DevConfig>): void {
    this.config = { ...this.config, ...newConfig };
    this.log('info', '开发配置已更新');
  }

  /**
   * 获取配置
   */
  public getConfig(): DevConfig {
    return { ...this.config };
  }
}